File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double myPow(double a, int b) {

        long n = b;
        double ans = 1.0;

        if (n < 0) {
            a = 1 / a;
            n = -n;
        }

        while (n > 0) {
            if (n % 2 == 1) {
                ans = ans * a;
            }
            a = a * a;
            n = n / 2;
        }

        return ans;
    }
}

File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    unordered_map<long long, double> dp;

    double pow(double x, long long n) {
        if (n == 0) return 1;

        if (dp.count(n)) return dp[n];  

        return dp[n] = x * pow(x, n - 1); 
    }

    double myPow(double x, int n) {
        long long N = n;
        if (N < 0) {
            x = 1.0 / x;
            N = -N;
        }
        return pow(x, N);
    }
};
